Flickr/Flickr/ImageCollectionViewController.swift

```diff
@@ -120,90 +158,217 @@ class ImageCollectionViewController: UIViewController,  UICollectionViewDelegate
+    }
+    
     @objc func editingTextEventFunc(textField: UITextField) {
-        searchHistoryHide()
+        if getFilteredHistory() {
+            return
+        } else {
+            startEditingEvent()
+        }
+    }
+    
+    @objc func refreshCurrentCollectionViewByPullToUpdate() {
+        print("reload")
+        pageFlickr = 1
+        isNotUpdating = true
+        if self.searchCollectionView.alpha == 0 {
+            self.performFlickrPopular()
+            return
+        } else {
+            guard let lastValue = self.lastEnteredTextValue else {
+                return
+            }
+            self.performFlickrSearch(url: lastValue)
+        }
     }
     
     @objc func clickOnTextEventFunc(textField: UITextField) {
-        searchHistoryView.reloadData()
-        searchHistoryShowMustGoOn()
+       _ = getFilteredHistory()
+        startEditingEvent()
+    }
+    
+    func isTextFieldEmpty() -> Bool {
+        guard let textField = searchTextField.text else {
+            return false
+        }
+        if textField.count == 0 || textField == "" {
+            return true
+        }
+        return false
     }
     
     func setDelegates_DataSources() {
         searchTextField.delegate = self
-        searchHistoryView.delegate = self
-        searchHistoryView.dataSource = self
+        searchHistoryTableView.delegate = self
+        searchHistoryTableView.dataSource = self
+    }
+    
+    func setMagnifyAndCancelButtonAlphasDefault() {
+        cancelButtonOutler.alpha = 0
+        magnifyImage.alpha = 0.5
     }
     
     func setAlphasDefault() {
         subViewForSpinner.alpha = 0
         searchCollectionView.alpha = 0
         popularCollectionView.alpha = 1
-        searchHistoryView.alpha = 0
+        searchHistoryTableView.alpha = 0
+        setMagnifyAndCancelButtonAlphasDefault()
+    }
+    
+    func scrollBack(view: UICollectionView, indexPath: IndexPath) {
+        view.scrollToItem(at: indexPath, at: .centeredVertically, animated: false)
     }
     
     func setConstraintMode() {
         searchConstraint.priority = UILayoutPriority(rawValue: 999)
         searchConstraint.isActive = true
     }
     
-    func setBehaviorTextField() {
+    @objc func tapImageAction(recognizer: UITapGestureRecognizer) {
+        searchTextField.becomeFirstResponder()
+        startEditingEvent()
+    }
+    
+    func startEditingEvent() {
+        UIView.animate(withDuration: 0.25,  animations: {
+            self.searchCollectionView.alpha = 0
+            self.popularCollectionView.alpha = 0
+            self.magnifyImage.alpha = 1
+            self.cancelButtonOutler.alpha = 1
+            self.searchHistoryTableView.alpha = 1
+            self.view.layoutIfNeeded()
+        })
+        searchHistoryTableView.reloadData()
+        rebuildTableSize()
+    }
+    
+    func isCollectionsViewHidedBoth() -> Bool {
+        let check = (searchCollectionView.alpha == 0) && (popularCollectionView.alpha == 0)
```

Comment:
using view properties for checking logic is wrong approach. Need make separate bool variable for visibility